File: dist_utils.py
import math;
import random;

# normalizes a distribution (as a dictionary from key to probability)
def normalize(D):
    N = dict();
    total = 0.0;
    for p in D.values():
        total = total + p;
    
    for k in D.keys():
        N[k] = D[k] / total;
    
    return N;

# creates a softmax distribution while avoiding overflow
def softmax_dist(D):
    a = max(D.values());
    Z = 0.0;
    
    P = dict();
    for k in D.keys():
        P[k] = math.exp(D[k] - a);
    
    return normalize(P);

# picks a random item from the (normalized) probability distribution
def random_item(D):
    r = random.random();
    for k in D.keys():
        r = r - D[k];
        if r <= 0:
            return k;
    
    raise Exception("ERROR: probability distribution sums to < 1.");

File: cpp/dist_utils.cpp
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <map>

#include "dist_utils.h"

using namespace std;

// normalizes a distribution (as a map from key to probability)
void normalize(map<int,double>& D) {
  double total = 0.0;
  
  map<int,double>::iterator it;
  for (it = D.begin(); it != D.end(); it++) {
    total += (*it).second;
  }
  for (it = D.begin(); it != D.end(); it++) {
    D[(*it).first] = (*it).second / total;
  }
}

// creates a softmax distribution while avoiding overflow
void softmax_dist(map<int,double>& D) {
  map<int,double>::iterator it;
  
  double a = 0.0;
  for (it = D.begin(); it != D.end(); it++) {
    a = max((*it).second, a);
  }
  
  for (it = D.begin(); it != D.end(); it++) {
    D[(*it).first] = exp((*it).second - a);
  }

  normalize(D);
}

// picks a random item from the (normalized) probability distribution
int random_item(map<int,double>& D) {
  double r = ((double)rand() / (double)RAND_MAX);
  
  map<int,double>::iterator it;
  for (it = D.begin(); it != D.end(); it++) {
    r -= (*it).second;
    if (r <= 0) return (*it).first;
  }
  cout << "ERROR: probability distribution sums to < 1." << endl;
  exit(1);
}
